File: examples/offline_signing/1_transaction_preparation.rs
// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! cargo run --example 1_transaction_preparation --release
use iota_client::{api::PreparedTransactionData, Client, Result};
use std::{
    fs::File,
    io::{prelude::*, BufWriter},
    path::Path,
};

/// In this example we will get inputs and prepare a transaction

const ADDRESS_FILE_NAME: &str = "examples/offline_signing/addresses.json";
const PREPARED_TRANSACTION_FILE_NAME: &str = "examples/offline_signing/prepared_transaction.json";

#[tokio::main]
async fn main() -> Result<()> {
    // Address to which we want to send the amount
    let address = "atoi1qruzprxum2934lr3p77t96pzlecxv8pjzvtjrzdcgh2f5exa22n6gek0qdq";
    let amount = 1_000_000;

    // Get inputs and create transaction essence online
    let iota_online = Client::builder()
        .with_node("https://api.lb-0.h.chrysalis-devnet.iota.cafe")? // Insert your node URL here
        .finish()
        .await?;

    let addresses = read_addresses_from_file(ADDRESS_FILE_NAME)?;

    let inputs = iota_online.find_inputs(addresses, amount).await?;

    // Prepare transaction
    let mut transaction_builder = iota_online.message();
    for input in inputs {
        transaction_builder = transaction_builder.with_input(input);
    }
    let prepared_transaction_data = transaction_builder
        .with_output(address, amount)?
        .prepare_transaction()
        .await?;

    println!("Prepared transaction sending {} to {}", amount, address);

    write_transaction_to_file(PREPARED_TRANSACTION_FILE_NAME, prepared_transaction_data)?;

    Ok(())
}

fn read_addresses_from_file<P: AsRef<Path>>(path: P) -> Result<Vec<String>> {
    let mut file = File::open(&path)?;
    let mut data = String::new();
    file.read_to_string(&mut data)?;
    let addresses: Vec<String> = serde_json::from_str(&data)?;
    Ok(addresses)
}

fn write_transaction_to_file<P: AsRef<Path>>(
    path: P,
    prepared_transaction_data: PreparedTransactionData,
) -> Result<()> {
    let jsonvalue = serde_json::to_value(&prepared_transaction_data)?;
    let file = File::create(path)?;
    let bw = BufWriter::new(file);
    serde_json::to_writer_pretty(bw, &jsonvalue)?;
    Ok(())
}

File: bindings/nodejs/examples/offline_signing/1_transaction_preparation.ts
// Copyright 2021-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0
import { Client, initLogger } from '@iota/client';
import { writeFile, readFile } from 'fs/promises';

// From examples directory, run with:
// node ./dist/offline_signing/1_transaction_preparation.js

// In this example we will get inputs and prepare a transaction
const ADDRESS_FILE_NAME = __dirname + '/../../offline_signing/addresses.json';
const PREPARED_TRANSACTION_FILE_NAME =
    __dirname + '/../../offline_signing/prepared_transaction.json';

async function run() {
    initLogger();

    // client will connect to testnet by default
    const onlineClient = new Client({
        nodes: [
            {
                // Insert your node URL here.
                url: 'http://localhost:14265/',
                disabled: false,
            },
        ],
        localPow: true,
    });

    const address =
        'rms1qqv5avetndkxzgr3jtrswdtz5ze6mag20s0jdqvzk4fwezve8q9vkpnqlqe';
    const amount = 1000000;
    try {
        // Recovers addresses from example `0_address_generation`.
        const addresses = JSON.parse(await readFile(ADDRESS_FILE_NAME, 'utf8'));

        // Gets enough inputs related to these addresses to cover the amount.
        const inputs = await onlineClient.findInputs(addresses, amount);

        // Prepares the transaction
        const preparedTransaction = await onlineClient.prepareTransaction(
            undefined,
            {
                inputs,
                output: { address, amount: amount.toString() },
            },
        );

        console.log(`Prepared transaction sending ${amount} to ${address}.`);

        await writeFile(
            PREPARED_TRANSACTION_FILE_NAME,
            JSON.stringify(preparedTransaction),
        );
    } catch (error) {
        console.error(error);
    }
}

run().then(() => process.exit());
